### PR TITLE
Ruby translation of GnuPG-checking Git hook

### PR DESCRIPTION
**Pull request by creideiki**
The Git server-side hook checking GnuPG signatures is currently (see https://github.com/saab-simc-admin/scripts) written in Bash, which makes me nervous. This is a straight translation into Ruby, using Rugged and GPGME.

This needs a lot more testing before it can be accepted.

### CODE REVIEW

File: git-hooks/enforce-workflow.rb [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    begin
+      signature, plaintext = Rugged::Commit.extract_signature(repo, commit.oid)
+      crypto.verify(signature, :signed_text => plaintext) do |signature|
+        signed = signature.valid?
+        next if not signed
+        signer = signature.fingerprint
+        allowed = is_allowed_signer(signer)
+      end
+    rescue Exception => e
+      puts e
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: Is it really safe to continue if there is an exception?

File: git-hooks/enforce-workflow.rb [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+        if not repo.config['hooks.allowunsignedtags'] or
+            not repo.config['hooks.allowunannotated']
+          puts "*** The un-annotated tag #{ref} is not allowed in this repository."
+          puts "*** Use 'git tag [ -a | -s ]' for tags you want to propagate."
+          exit 1
+        end
+      when :tag
+        # The ref is an annotated tag
+        if rev_old.to_i(16) == 0 and not repo.config['hooks.allowunsignedtags']
+          puts "*** Tag #{ref} already exists."
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: The doc at top says that the config options for this is `hooks.allowmodifytag`

File: git-hooks/enforce-workflow.rb [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+          puts "*** Modifying a tag is not allowed in this repository."
+        else
+          if not repo.config['hooks.allowunsignedtags']
+            if allowed
+              puts "*** Good signature on tag #{ref} by #{signer}."
+            else
+              puts "*** Rejecting tag #{ref} due to lack of a valid GPG signature."
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: Harmonize the signature rejection handling for tags and commits.

File: git-hooks/enforce-workflow.rb [hunk 4]
```diff
@@ -0,0 +1,236 @@
...
+#
+# where <old-value> is the old object name stored in the ref,
+# <new-value> is the new object name to be stored in the ref and
+# <ref-name> is the full name of the ref. When creating a new ref,
+# <old-value> is 40 0.
+
+STDIN.each do |line|
+  rev_old, rev_new, ref = line.split
+
+  if not repo.config['hooks.allowunsignedtags']
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: The top comment states that the config options for this is `hooks.allowcommitsonmaster`.

File: git-hooks/enforce-workflow.rb [hunk 5]
```diff
@@ -0,0 +1,236 @@
...
+    ##
+    ## We have to check this in a separate step, since we won't have
+    ## enough context while walking through the commit list to do this
+    ## check properly.
+    if ref == "refs/heads/master"
+      # We only want merges on master. A merge is a commit with at
+      # least two parents, and one of them has to be the old target.
+      parents = repo.lookup(rev_new).parent_ids
+      if parents.length < 2 or
+          not parents.include?(rev_old)
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: The initial commit on `master` also needs to be pushable.

File: git-hooks/enforce-workflow.rb [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+require 'gpgme'
+require 'yaml'
+
+repo = Rugged::Repository.new('.')
+crypto = GPGME::Crypto.new
+
+@collaborators = YAML.load_file(repo.path + 'collaborators.yaml')
+
+def is_allowed_signer(keyid)
+  keys = GPGME::Key.find(keyid)
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: `GPGME::Key.find` does not work that way. The first arg should be `:public` to find public keys, and the needle should be in the second arg.

File: git-hooks/enforce-workflow.rb [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+        puts "*** Master only accepts merges of feature branches."
+        exit 1
+      end
+    end
+  end
+
+  walker = Rugged::Walker.new(repo)
+
+  # Get all new commits on branch ref, even if it's a new branch.
+  walker.push(rev_new)
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: When pushing a branch deletion, `rev_new` is zeroes, which cannot be pushed.

File: git-hooks/enforce-workflow.rb [hunk 8]
```diff
@@ -0,0 +1,236 @@
...
+  walker.each do |commit|
+    # walker.count would have consumed the walker, so instead track
+    # the number of commits manually.
+    commit_count += 1
+
+    puts commit.oid
+
+    if commit.oid.to_i(16) == 0
+      # A commit with a hash full of zeros is a deletion of a ref.
+      commit_type = :delete
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: I don't think ref deletions can appear in the actual commit graph.

File: git-hooks/enforce-workflow.rb [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+        exit 1
+      elsif ref =~ /^refs\/tags\// and not repo.config['hooks.allowdeletetag']
+        puts "*** Deleting a tag is not allowed in this repository."
+        exit 1
+      end
+    else
+      puts "*** Unknown type of update to ref #{ref} of type #{commit_type}."
+    end
+
+    if commit_count == 0
```
Comment (notCalle) [inline git-hooks/enforce-workflow.rb]: This condition is always false unless moved outside the `walker.each` block.

**Review by notCalle — CHANGES_REQUESTED** [review]
In general very difficult to read, because there is a `.each do ... end` spanning > 100 lines that does almost everything in deeply nested logic structures.

The overall design is insecure because it accepts by default. The default should be to reject, unless the specific circumstances where a push is allowed has been verified.

Probably needs a `Gemfile` to describe its dependencies.

**Comment by creideiki** [comment]
Pushed 9045ea1 which I think should solve all actual bugs before I start restructuring the logic. However, the call to `Rugged::Commit.extract_signature` likes to corrupt the last couple of bytes of the plaintext on my system, so it makes spurious claims of signatures being bad.

**Comment by creideiki** [comment]
Signature validation fails due to a bug in libgit2: libgit2/libgit2#4118

**Comment by creideiki** [comment]
I no longer see random signature validation failures as of Rugged 0.26.0b5.

**Comment by creideiki** [comment]
This should be ready for another look now.

**Comment by notCalle** [comment]
Pushing the initial commit to an empty repo still fails.
```
remote: *** Accepting creation of master branch.
To /Users/calle/Workspace/signed-only.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to '/Users/calle/Workspace/signed-only.git'
```

**Comment by creideiki** [comment]
I have no idea what you are talking about. It's always worked wonderfully, with no bugs whatsoever.

**Comment by notCalle** [comment]
Using `git` >= 2.11 seems to break `pre-receive`, so that should be noted somewhere... 😞

`git` 2.11.0 and 2.12.2:

```
Counting objects: 3, done.
Writing objects: 100% (3/3), 845 bytes | 0 bytes/s, done.
Total 3 (delta 0), reused 0 (delta 0)
remote: hooks/pre-receive:126:in `push': object not found - no match for id (f5f1e8e9f4511a7c83de6eefb0b99c9b5331e03a) (Rugged::OdbError)
remote: 	from hooks/pre-receive:126:in `walk_graph'
remote: 	from hooks/pre-receive:288:in `block in <main>'
remote: 	from hooks/pre-receive:266:in `each'
remote: 	from hooks/pre-receive:266:in `<main>'
remote: *** Accepting creation of master branch.
To /Users/calle/Workspace/signed-only.git
 ! [remote rejected] master -> master (pre-receive hook declined)
error: failed to push some refs to '/Users/calle/Workspace/signed-only.git'
```

`git` 2.10.1:

```
Counting objects: 3, done.
Writing objects: 100% (3/3), 845 bytes | 0 bytes/s, done.
Total 3 (delta 0), reused 0 (delta 0)
remote: *** Accepting creation of master branch.
remote: *** Good signature on commit f5f1e8e9f4511a7c83de6eefb0b99c9b5331e03a by calle@discord.bofh.se (1764CE72729D4C843AF6999A78D9BF4A13F75F59).
To /Users/calle/Workspace/signed-only.git
 * [new branch]      master -> master
```

**Comment by notCalle** [comment]
Pushing a new (signed) tag fails:
```
Counting objects: 1, done.
Writing objects: 100% (1/1), 793 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
remote: *** Tag refs/tags/with-readme already exists.
remote: *** Modifying a tag is not allowed in this repository.
To /Users/calle/Workspace/signed-only.git
 ! [remote rejected] with-readme -> with-readme (pre-receive hook declined)
error: failed to push some refs to '/Users/calle/Workspace/signed-only.git'
```

**Comment by notCalle** [comment]
Pushing a new unsigned tag fails with stack trace instead of proper error message.
```
Counting objects: 1, done.
Writing objects: 100% (1/1), 181 bytes | 0 bytes/s, done.
Total 1 (delta 0), reused 0 (delta 0)
remote: /usr/local/lib/ruby/gems/2.4.0/gems/gpgme-2.0.12/lib/gpgme/ctx.rb:447:in `verify': No data (GPGME::Error::NoData)
remote: 	from /usr/local/lib/ruby/gems/2.4.0/gems/gpgme-2.0.12/lib/gpgme/crypto.rb:312:in `block in verify'
remote: 	from /usr/local/lib/ruby/gems/2.4.0/gems/gpgme-2.0.12/lib/gpgme/ctx.rb:79:in `new'
remote: 	from /usr/local/lib/ruby/gems/2.4.0/gems/gpgme-2.0.12/lib/gpgme/crypto.rb:311:in `verify'
remote: 	from hooks/pre-receive:237:in `allow_annotated_tag?'
remote: 	from hooks/pre-receive:308:in `block in <main>'
remote: 	from hooks/pre-receive:270:in `each'
remote: 	from hooks/pre-receive:270:in `<main>'
To /Users/calle/Workspace/signed-only.git
 ! [remote rejected] with-readme -> with-readme (pre-receive hook declined)
error: failed to push some refs to '/Users/calle/Workspace/signed-only.git'
```

File: git-hooks/enforce-workflow/enforce-workflow.rb [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+      end
+    end
+    if !signed
+      puts "*** Bad signature on tag #{ref}."
+      return false
+    elsif signer
+      puts "*** Good signature on tag #{ref} by #{signer} (#{fingerprint})."
+      return true
+    else
+      puts "*** Rejecting tag #{ref} due to lack of a valid GPG signature."
```
Comment (notCalle) [inline git-hooks/enforce-workflow/enforce-workflow.rb]: Lies! It has a valid signature, but the signer is not authorized (see `walk_graph`). 

Also, maybe consider a DRYer approach to signature validation and reporting?

**Review by notCalle — APPROVED** [review]

**Pull request merged**